apps/OboeTester/app/src/main/cpp/AudioStreamGateway.h

```diff
@@ -19,28 +19,24 @@
 #include <unistd.h>
 #include <sys/types.h>
 
-#include "AudioProcessorBase.h"
+#include "flowgraph/AudioProcessorBase.h"
 #include "oboe/Oboe.h"
 
+using namespace flowgraph;
+
 /**
  * Bridge between an audio graph and an audio device.
- * Connect the audio units to the "input" and then pass
+ * Pass in an AudioSink and then pass
  * this object to the AudioStreamBuilder as a callback.
  */
-class AudioStreamGateway : public AudioProcessorBase, public oboe::AudioStreamCallback {
+class AudioStreamGateway : public oboe::AudioStreamCallback {
 public:
     AudioStreamGateway(int samplesPerFrame);
     virtual ~AudioStreamGateway();
 
-    /**
-     * Process audio for the graph.
-     * @param framePosition
-     * @param numFrames
-     * @return
-     */
-    AudioResult onProcess(
-            uint64_t framePosition,
-            int numFrames) override;
+    void setAudioSink(flowgraph::AudioSink  *sink) {
```

Comment:
II am trying to decide whether to just declare flowgraph non-thread-safe or to take the performance hit from the shared pointers.

Reply:
I realize that if I use shared_ptr then I might cause the real-time audio thread to delete an object. That could cause a glitch. So I am inclined to document that making or breaking connections and deleting graph nodes  is not thread-safe.

Reply:
Regarding performance, I thought there was only a performance hit when the shared_ptr ref count was changed (since it's an atomic operation), in your case isn't the flow graph largely static after setup so the performance hit would be minimal?

Regarding object deletion, couldn't you have a mechanism which marks a node for deletion and then some process (either the audio thread or the owning thread) deletes it only when it is safe to do so? I know that's sounding suspiciously like garbage collection.

Reply:
If the audio thread does the deletion then it will glitch.
There is never a "safe" time for the callback thread to do deletions, even if they are deferred.

This patch did not create the problem of dynamic deletion. And it is not meant to be the final patch that makes this perfect. I think we should defer this issue to a later CL. I would like to go ahead and merge this after addressing issues in this patch.